### PR TITLE
[MPS] Clear MPSGraphCache in torch.mps.empty_cache()

### LINKED ISSUE #181213: [MPS] Unbounded process RSS growth with varying-shape inference on macOS
# [MPS] Unbounded process RSS growth during inference with attention-heavy models on macOS

## Summary

On macOS with Apple Silicon, repeated inference forward passes through an attention-heavy model on the MPS device cause process RSS to grow linearly with the number of calls, while `torch.mps.current_allocated_memory()`, `torch.mps.driver_allocated_memory()`, and the count of live MPS tensors visible to `gc.get_objects()` all stay flat. The leak is invisible to PyTorch's memory accounting and cannot be reclaimed by any public API (`empty_cache`, `synchronize`, `inference_mode`, moving the model off MPS and back, NSAutoreleasePool push/pop, etc.).

## Environment

- PyTorch: **2.10.0**
- OS: macOS on Apple Silicon
- Device: `mps`

## Reproducer

Attached: [`repro.py`](./repro.py). Self-contained, no dependencies beyond `torch`. Runs a stack of attention blocks (alternating pure-matmul manual attention and `nn.MultiheadAttention`) in an inference loop and reports RSS alongside `torch.mps` counters.

Two modes:

```bash
# Reproduces the leak on MPS:
python3 repro.py --device mps --shape-mode varying --calls 500

# Controls (all three should plateau):
python3 repro.py --device mps --shape-mode fixed   --calls 500
python3 repro.py --device cpu --shape-mode varying --calls 500
python3 repro.py --device cpu --shape-mode fixed   --calls 500
```

**The leak is triggered specifically by varying intermediate tensor shapes between calls, not by attention alone.** In `fixed` mode the input sequence length is constant; in `varying` mode each call samples a new length in `[seq_len // 8, seq_len]`.

### Reproducer results (PyTorch 2.10.0, macOS Apple Silicon)

Summary of `Δrss` (delta from call-1 baseline) across the four configurations:

| Config             | call 50 | call 100 | call 250 | call 500 | Behavior |
|--------------------|--------:|---------:|---------:|---------:|----------|
| **mps + varying**  | +74.7 MB | +145.7 MB | **+314.3 MB** | **+556.8 MB** | **monotonic leak** (~1 MB/call) |
| mps + fixed        |  +0.1 MB |   +0.1 MB |   +0.3 MB |   +0.4 MB | plateaus |
| cpu + varying      | +271.2 MB | +482.4 MB | +491.9 MB | +493.9 MB | **plateaus** after ~100 calls |
| cpu + fixed        |  +8.6 MB |  +15.6 MB |  +15.6 MB |  +15.6 MB | plateaus |

The CPU + varying case is the critical control: it *also* sees the full range of intermediate shapes, and its allocator pool grows in the first ~100 calls to hold them, then **fully plateaus** (stays at 491.9 MB for calls 150 → 500). That is the behavior of a bounded cache being reused. MPS + varying adds ~50 MB per 50 calls indefinitely over the same workload.

Throughout the MPS + varying run, `torch.mps.current_allocated_memory()` stays flat at ~14 MB, `torch.mps.driver_allocated_memory()` oscillates within 136-148 MB with no trend, and the live MPS tensor count from `gc.get_objects()` stays flat at 50. PyTorch's own memory accounting is completely blind to the ~557 MB that has been added to process RSS.

Full transcripts in `<details>` below.

<details>
<summary>mps + varying (the leak)</summary>

```
PyTorch:    2.10.0
Device:     mps
Shape mode: varying (seq_len in [128, 1024])
Stack:      4 blocks × 8 heads × d_model=256
  call     rss_mb       Δrss    mps_cur    mps_drv   live_t
----------------------------------------------------------------------
     1      273.9       +0.0       14.0       88.7       50
    50      348.6      +74.7       14.0      141.4       50
   100      419.6     +145.7       13.7      144.2       50
   150      478.0     +204.1       14.0      136.9       50
   200      539.0     +265.1       13.8      145.6       50
   250      588.1     +314.3       13.8      146.1       50
   300      642.1     +368.2       13.9      146.7       50
   350      691.3     +417.5       13.5      139.2       50
   400      734.8     +461.0       13.5      139.8       50
   450      780.0     +506.2       13.8      148.3       50
   500      830.6     +556.8       14.0      148.8       50
Total elapsed: 7.0s  (71.0 calls/s)
```

</details>

<details>
<summary>mps + fixed (control)</summary>

```
PyTorch:    2.10.0
Device:     mps
Shape mode: fixed (seq_len = 1024)
Stack:      4 blocks × 8 heads × d_model=256
  call     rss_mb       Δrss    mps_cur    mps_drv   live_t
----------------------------------------------------------------------
     1      273.6       +0.0       14.1       90.7       50
    50      273.7       +0.1       14.1       90.7       50
   100      273.7       +0.1       14.1       90.7       50
   150      273.8       +0.1       14.1       90.7       50
   200      273.9       +0.2       14.1       90.7       50
   250      273.9       +0.3       14.1       90.7       50
   300      273.9       +0.3       14.1       90.7       50
   350      273.9       +0.3       14.1       90.7       50
   400      274.0       +0.3       14.1       90.7       50
   450      274.0       +0.4       14.1       90.7       50
   500      274.0       +0.4       14.1       90.7       50
Total elapsed: 3.5s  (142.5 calls/s)
```

</details>

<details>
<summary>cpu + varying (control — bounded cache)</summary>

```
PyTorch:    2.10.0
Device:     cpu
Shape mode: varying (seq_len in [128, 1024])
Stack:      4 blocks × 8 heads × d_model=256
  call     rss_mb       Δrss    mps_cur    mps_drv   live_t
----------------------------------------------------------------------
     1      291.8       +0.0        0.0        0.4       50
    50      563.0     +271.2        0.0        0.4       50
   100      774.2     +482.4        0.0        0.4       50
   150      782.6     +490.7        0.0        0.4       50
   200      783.7     +491.9        0.0        0.4       50
   250      783.7     +491.9        0.0        0.4       50
   300      783.7     +491.9        0.0        0.4       50
   350      783.8     +491.9        0.0        0.4       50
   400      783.8     +491.9        0.0        0.4       50
   450      783.8     +491.9        0.0        0.4       50
   500      785.8     +493.9        0.0        0.4       50
Total elapsed: 7.2s  (69.6 calls/s)
```

</details>

<details>
<summary>cpu + fixed (control)</summary>

```
PyTorch:    2.10.0
Device:     cpu
Shape mode: fixed (seq_len = 1024)
Stack:      4 blocks × 8 heads × d_model=256
  call     rss_mb       Δrss    mps_cur    mps_drv   live_t
----------------------------------------------------------------------
     1      296.5       +0.0        0.0        0.4       50
    50      305.0       +8.6        0.0        0.4       50
   500      312.1      +15.6        0.0        0.4       50
Total elapsed: 13.4s  (37.2 calls/s)
```

</details>

(An initial version of this reproducer used fixed shapes only and failed to reproduce the leak despite the real-world model showing 4 MB/call of RSS growth. That was the diagnostic clue that pointed at shape-keyed cache growth rather than a per-call allocator bug.)

## Observed symptom fingerprint

From a real-world model (drone navigation policy; PointTransformerV3-style encoder + MultiheadAttention-based nav head) with 50 forward calls per "episode", no renderer, fixed-shape inputs:

```
[MEM ep1] dev=mps RSS:  874.2 MB   mps_cur: 32.2 MB   mps_drv:  63.0 MB   mps_tensors: 285
[MEM ep2] dev=mps RSS: 1041.0 MB   mps_cur: 32.2 MB   mps_drv:  53.6 MB   mps_tensors: 285
[MEM ep3] dev=mps RSS: 1263.2 MB   mps_cur: 32.2 MB   mps_drv:  54.4 MB   mps_tensors: 285
[MEM ep4] dev=mps RSS: 1462.9 MB   mps_cur: 32.2 MB   mps_drv:  65.2 MB   mps_tensors: 285
[MEM ep5] dev=mps RSS: 1730.4 MB   mps_cur: 32.2 MB   mps_drv:  60.4 MB   mps_tensors: 285
```

RSS grows ~200 MB per 50 calls (~4 MB/call) while `current_allocated`, `driver_allocated`, and MPS tensor count are all flat.

## Bisection — which ops leak?

Per-call RSS delta measured at ~10 boundaries inside the forward pass, averaged over 50 calls:

| Forward segment | avg RSS delta / call | Notes |
|---|---|---|
| `ptv3_encoder` call | **4086 KB** | Manual `Q @ Kᵀ → softmax → @V`, no `nn.MultiheadAttention`, no SDPA |
| nav head scene cross-attn (`nn.MultiheadAttention` + LN + FFN) | **1337 KB** | |
| nav head memory-token construction + concat | 622 KB | Small attention surface, pos_enc + concat |
| `ptv3_fusion` (2 × `nn.TransformerEncoderLayer`) | 242 KB | |
| All other segments combined | < 130 KB | noise |

**Key observation:** both the manual-matmul attention (PTv3's `SerializedAttentionBlock`) and the `nn.MultiheadAttention`-based path leak. The per-call magnitude scales with the attention compute volume, not with any specific API. This rules out a bug in `nn.MultiheadAttention`'s fast path and points at something common to all attention-shaped compute graphs on MPS.

Input shapes are fixed (padded to `max_objects=40`, `n_points=256` per object) — not a cache-churn-from-shape-variation issue.

## What's been ruled out

- **Renderer / GPU graphics libraries.** Leak reproduces identically with the 3D renderer disabled; reproduces across two different rendering backends (VTK/PyVista and Open3D/Filament).
- **PyTorch tensor accumulation.** `torch.mps.current_allocated_memory()` flat. Live MPS tensor count from `gc.get_objects()` flat at 285 across episodes.
- **Python heap leak.** Object counts by type via `gc.get_objects()` are flat episode-over-episode (same `function:+N, tuple:+N, dict:+N, ...` deltas every time).
- **Autograd graph retention.** Forward pass wrapped in `@torch.inference_mode()`; no effect.
- **MPS caching allocator pool.** `torch.mps.empty_cache()` called per step and per episode; `PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0` set; no effect.
- **Metal command queue retention.** `torch.mps.synchronize()` called after every forward pass; no effect.
- **NSAutoreleasePool accumulation.** Manual `objc_autoreleasePoolPush` / `objc_autoreleasePoolPop` via `ctypes` wrapped around each inference call; partial (~30% reduction) but does not eliminate the leak.
- **Model-level state.** In the real-world model, a `model.to('cpu') → empty_cache → model.to('mps')` cycle every N episodes **did not reduce RSS**, and actually increased it slightly (the subsequent forward calls pay the cost of re-compiling `MPSGraph` entries for the shapes seen on the new device context). Consistent with the cache being process-global rather than model-bound.
- **Old PyTorch.** Reproduced on PyTorch 2.10.0.

## Hypothesis

The retention fingerprint:
- grows per call, magnitude proportional to op compute volume
- triggered by varying intermediate tensor shapes, not by attention itself
- invisible to `torch.mps.current_allocated_memory()` and `torch.mps.driver_allocated_memory()`
- device-level rather than model-level (moving the model off MPS and back *increases* the leak)
- unaffected by `empty_cache`, `synchronize`, `inference_mode`, `NSAutoreleasePool`

…is consistent with **unbounded growth in Metal's MPSGraph compilation cache, keyed on intermediate tensor shapes, with no eviction path**. Each distinct intermediate shape the forward pass produces (grid-pool voxel counts, memory-cache token counts, PTv3 windowed sequence lengths after padding, etc.) appears to produce a new `MPSGraph` entry that is then retained for the lifetime of the process.

The reproducer above isolates this: attention with identical shapes every call plateaus; attention with shapes that resample per call leaks. This strongly suggests the issue is in PyTorch's MPSGraph caching layer rather than in any individual op.

## Ask

- Confirm whether this is a known issue / tracked internally.
- Guidance on any non-public API that can evict the MPSGraph / pipeline cache at runtime (to let users bound the leak without process restart).
- If not already tracked, a fix targeting the attention path (both `nn.MultiheadAttention` and manual matmul + softmax).

Happy to instrument further or run diagnostic builds.

---

Additional diagnostic output and the full bisection transcript available on request.

cc @jerryzh168 @malfet @aditvenk @kulinseth @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
## Summary

`MPSGraphCache` is an unbounded singleton cache storing compiled `MPSGraph` objects keyed by tensor shape/dtype/operation signatures. For workloads with variable tensor shapes (e.g. GNN training with variable-size graphs, NLP with variable-length sequences, or any dynamic-shape workload), every unique shape produces a cache entry that may never be reused, causing Metal driver memory to grow without bound.

`torch.mps.empty_cache()` only cleared the MPS allocator pool, leaving the graph cache untouched — users had no way to reclaim the leaked memory short of killing the process.

**This change:**
1. Adds a `clear()` method to `MPSGraphCache` that safely deletes all cached graphs under the serial queue lock
2. Wires `MPSHooks::emptyCache()` to call `MPSGraphCache::clear()` alongside the existing allocator clear

Fixes #181213, #132596
Related: #164299, #152550

## Test plan

Tested on Apple M4 Max (macOS 15) with a variable-shape GNN training workload (batch_size=32, 315 batches/epoch where each batch has a unique graph topology).

| Stage | MPS driver memory |
|---|---|
| Before training | 9 MB |
| After 1 epoch (315 unique shapes) | 13,692 MB |
| After `torch.mps.empty_cache()` | **52 MB** |
| After epoch 2 | 14,730 MB |
| After 2nd `empty_cache()` | **52 MB** |

Memory measured via `torch.mps.driver_allocated_memory()`. The allocator-tracked memory (`current_allocated_memory()`) remained ~10 MB throughout — confirming the leak is in the graph cache, not the tensor allocator.

**Before this fix:** MPS driver memory grew ~14 GB/epoch with no reclaim path.
**After this fix:** `torch.mps.empty_cache()` drops driver memory from ~14 GB to 52 MB.

cc @malfet @kulinseth

### CODE REVIEW

**Comment by Skylion007** [comment]
Would just setting a maximum cache size and having it be an LRU be more efficient I wonder?

**Comment by anagnorisis2peripeteia** [comment]
That's a great point. An LRU would definitely be the cleaner solution for workloads with repeating tensor shapes where the cache hits actually matter.

In my case (GNN training with variable-size graphs), each batch has a unique shape that's never reused, so the cache entries are essentially dead weight — but I appreciate that's not the common case.

This PR is mainly just giving `empty_cache()` the ability to clear the graph cache at all, since it currently can't. It's fully opt-in and doesn't change default behaviour. I think it'd be needed even with an LRU, since you'd still want a way to flush on demand.

Would be interested to hear what the MPS maintainers think about the right approach here.

**Review by malfet — APPROVED** [review]
This would definitely work, though I agree with @Skylion007 that having some sort of thresholds would be better and/or replacing MPSGraph calls with the Metal shaders

File: aten/src/ATen/native/mps/OperationUtils.h [hunk 1]
```diff
@@ -389,6 +389,15 @@ struct MPSGraphCache {
     return static_cast<T*>(LookUp(key));
   }
 
+  void clear() {
+    dispatch_sync(serialQueue_, ^() {
+      for (const auto& i : cache_) {
+        delete i.second.cachedGraph_;
```
Comment (malfet) [inline aten/src/ATen/native/mps/OperationUtils.h:395]: This will open up a potential for use-after-free if other thread is using the entry returned by `LookUpOrCreate` function call

Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/OperationUtils.h:395]: Thanks for flagging this — I believe the serial dispatch queue (`serialQueue_`) already handles the thread safety here. `clear()` dispatches through the same serial queue as `LookUp()` and `CreateCachedGraph()`, so they can't execute concurrently. Happy to add a mutex if you think it's warranted though — just wanted to check whether the existing serialization is sufficient?

**Comment by anagnorisis2peripeteia** [comment]
The `crossref` failure is a 210-minute infrastructure timeout unrelated to this PR. The actual errors in the log are docstring parsing issues in `torch.sparse`, `torch.distributed`, and similar modules — none of which are touched by this change. The `empty_cache()` fix is purely a C++ call to `[MPSGraphCache clearCache]` with no Python-visible API surface. Could someone trigger a re-run of the crossref job?
